### DISCUSSION TITLE
Appending to template `toJson`

### DISCUSSION THREAD

**Discussion by LecrisUT** [smallstep/certificates#546] [Q&A] [answered]
I have looked into the `text/template` and `sprig` documentation and tried a few options, but I can't find how to add additional variables to the `toJson` part. For example, I want to setup:
```
   	"principals": {{ append .Principals .Token.preferred_username | toJson }},
```
Also using `toJson .Token.preferred_username` fails because it's not an array, but is there a way to test/force it to be a Json array?

**Comment by maraino** [comment, marked as answer]
The line that you have should work if `.Token.preferred_username` is a string.

If you want to use just that you can use `"principals": {{ list .Token.preferred_username | toJson }}`

**Reply by LecrisUT** [reply]
That would resolve `{{ toJson .Token.preferred_username }}`, but would it work as `{{ list .Principals .Token.preferred_username }}`?

Edit: It doesn't work:
```
 error unmarshaling certificate: json: cannot unmarshal array into Go struct field Certificate.principals of type string
```

**Reply by maraino** [reply]
The example that you had was working for me without issues:
```
"principals": {{ append .Principals .Token.preferred_username | toJson }},
```

**Reply by maraino** [reply]
I've created a simple program to show this working https://github.com/maraino/ssh-template-test

It outputs a certificate with the `foobar` principal coming from the token:
```
{
  "Nonce": "TE4ke1ZETVxuWnAzKVNARzVzeH1uc25jPjsibmJKbzk=",
  "Key": "1RAl/m05whrOLQWml4Hb3mrmmsJWL7nduAoflgEQGMM=",
  "Serial": 17283346687152004381,
  "CertType": 1,
  "KeyId": "mariano@smallstep.com",
  "ValidPrincipals": [
    "mariano",
    "foobar"
  ],
  "ValidAfter": 0,
  "ValidBefore": 0,
  "CriticalOptions": null,
  "Extensions": {
    "permit-X11-forwarding": "",
    "permit-agent-forwarding": "",
    "permit-port-forwarding": "",
    "permit-pty": "",
    "permit-user-rc": ""
  },
  "Reserved": null,
  "SignatureKey": "th+PJQE5AEblE9NacOZrsVMtsuK77l5oc3/X7Pcz3Z4=",
  "Signature": {
    "Format": "ssh-ed25519",
    "Blob": "oyeoD4v7GXZ7VuiViRJ/WtPvrXwLoNZ9YKFCbTTkg4CCXFRFy3LWryhOypy4/WSNfX3JxYcHC9bvSEDRFRSyAQ==",
    "Rest": null
  }
}
````

**Reply by LecrisUT** [reply]
You are right. I must have had another error in the template when I first tested this. Sorry for the confusion.
